pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,147 @@
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	interval = 1 * time.Minute
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+	// WaitGroup for waiting all deployment chain updaters to be completed.
+	wg sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	d.logger.Info("start running DeploymentChainController")
+
+	t := time.NewTicker(interval)
+	defer t.Stop()
+	d.logger.Info("start syncing updaters")
+
+	for {
+		select {
+		case <-ctx.Done():
+			d.logger.Info("deploymentChainController has been stopped")
+			return nil
+
+		case <-t.C:
+			start := time.Now()
+			if err := d.syncUpdaters(ctx); err != nil {
+				d.logger.Error("failed to sync deployment chains", zap.Error(err))
+				continue
+			}
+			d.logger.Info("successfully sync all deployment chains", zap.Duration("duration", time.Since(start)))
+		}
+	}
+}
+
+func (d *DeploymentChainController) syncUpdaters(ctx context.Context) error {
+	// Remove done updater of completed deployment chain.
+	for id, u := range d.updaters {
+		if u.IsDone() {
+			d.logger.Info("remove done updater of deployment chain",
+				zap.String("id", id),
+				zap.Time("completed_at", u.doneTimestamp),
+			)
+			delete(d.updaters, id)
+		}
+	}
+
+	// Find all not completed deployment chains and create updater if does not exist.
+	notCompletedChains, err := listNotCompletedDeploymentChain(ctx, d.deploymentChainStore)
+	if err != nil {
+		return err
+	}
+	for _, c := range notCompletedChains {
+		if _, ok := d.updaters[c.Id]; !ok {
+			d.updaters[c.Id] = newUpdater(
+				c,
+				d.applicationStore,
+				d.deploymentStore,
+				d.deploymentChainStore,
+				d.logger,
+			)
+		}
+	}
+
+	d.logger.Info(fmt.Sprintf("there are %d running deployment chain updaters", len(d.updaters)))
+
+	for chainID := range d.updaters {
+		updater := d.updaters[chainID]
+		d.wg.Add(1)
+		go func() {
+			defer d.wg.Done()
+			if err := updater.Run(ctx); err != nil {
+				d.logger.Error("failed to update deployment chain",
+					zap.String("deploymentChainId", updater.deploymentChainID),
+					zap.Error(err),
+				)
+			}
+		}()
+	}
```

Comment:
Let's apply the worker pattern to avoid spawning an unpredictable amount of goroutines.
For example, we can have `max(len(updaters), 10)`  workers to handle the updater pool.